<!doctype html><html lang=en><head><meta charset=utf-8><meta http-equiv=x-ua-compatible content="IE=edge,chrome=1"><title>Tidy-transcriptomics manifesto - The Tidyomics Blog</title><meta name=renderer content="webkit"><meta name=viewport content="width=device-width,initial-scale=1,maximum-scale=1"><meta http-equiv=cache-control content="no-transform"><meta http-equiv=cache-control content="no-siteapp"><meta name=theme-color content="#f8f5ec"><meta name=msapplication-navbutton-color content="#f8f5ec"><meta name=apple-mobile-web-app-capable content="yes"><meta name=apple-mobile-web-app-status-bar-style content="#f8f5ec"><meta name=author content="Stefano & Maria"><meta name=description content=" "><meta name=keywords content="Hugo,theme,even"><meta name=generator content="Hugo 0.119.0 with theme even"><link rel=canonical href=https://tidyomics.github.io/tidyomicsBlog/post/2021-07-07-tidy-transcriptomics-manifesto/><link rel=apple-touch-icon sizes=180x180 href=/tidyomicsBlog/apple-touch-icon.png><link rel=icon type=image/png sizes=32x32 href=/tidyomicsBlog/favicon-32x32.png><link rel=icon type=image/png sizes=16x16 href=/tidyomicsBlog/favicon-16x16.png><link rel=manifest href=/tidyomicsBlog/manifest.json><link rel=mask-icon href=/tidyomicsBlog/safari-pinned-tab.svg color=#5bbad5><link href=/tidyomicsBlog/sass/main.min.7063e6a5193600da37b11a3f96b28adfb25747a97881e5f3aa057b41563749a4.css rel=stylesheet><link rel=stylesheet href=https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@3.1.20/dist/jquery.fancybox.min.css integrity="sha256-7TyXnr2YU040zfSP+rEcz29ggW4j56/ujTPwjMzyqFY=" crossorigin=anonymous><meta property="og:title" content="Tidy-transcriptomics manifesto"><meta property="og:description" content="






"><meta property="og:type" content="article"><meta property="og:url" content="https://tidyomics.github.io/tidyomicsBlog/post/2021-07-07-tidy-transcriptomics-manifesto/"><meta property="article:section" content="post"><meta property="article:published_time" content="2021-07-07T00:00:00+00:00"><meta property="article:modified_time" content="2021-07-07T14:16:36+10:00"><meta itemprop=name content="Tidy-transcriptomics manifesto"><meta itemprop=description content="






"><meta itemprop=datePublished content="2021-07-07T00:00:00+00:00"><meta itemprop=dateModified content="2021-07-07T14:16:36+10:00"><meta itemprop=wordCount content="2913"><meta itemprop=keywords content="tidyomics/tidyomicsBlog,manifesto,transcriptomics,tidyverse,"><meta name=twitter:card content="summary"><meta name=twitter:title content="Tidy-transcriptomics manifesto"><meta name=twitter:description content="






"><!--[if lte IE 9]><script src=https://cdnjs.cloudflare.com/ajax/libs/classlist/1.1.20170427/classList.min.js></script><![endif]--><!--[if lt IE 9]><script src=https://cdn.jsdelivr.net/npm/html5shiv@3.7.3/dist/html5shiv.min.js></script>
<script src=https://cdn.jsdelivr.net/npm/respond.js@1.4.2/dest/respond.min.js></script><![endif]--></head><body><div id=mobile-navbar class=mobile-navbar><div class=mobile-header-logo><a href=/ class=logo>The Tidyomics Blog</a></div><div class=mobile-navbar-icon><span></span>
<span></span>
<span></span></div></div><nav id=mobile-menu class="mobile-menu slideout-menu"><ul class=mobile-menu-list><a href=/tidyomicsBlog/><li class=mobile-menu-item>Home</li></a><a href=/tidyomicsBlog/post/><li class=mobile-menu-item>Archives</li></a><a href=/tidyomicsBlog/tags/><li class=mobile-menu-item>Tags</li></a><a href=/tidyomicsBlog/categories/><li class=mobile-menu-item>Categories</li></a><a href=/tidyomicsBlog/about/><li class=mobile-menu-item>About</li></a></ul></nav><div class=container id=mobile-panel><header id=header class=header><div class=logo-wrapper><a href=/tidyomicsBlog/ class=logo>The Tidyomics Blog</a></div><nav class=site-navbar><ul id=menu class=menu><li class=menu-item><a class=menu-item-link href=/tidyomicsBlog/>Home</a></li><li class=menu-item><a class=menu-item-link href=/tidyomicsBlog/post/>Archives</a></li><li class=menu-item><a class=menu-item-link href=/tidyomicsBlog/tags/>Tags</a></li><li class=menu-item><a class=menu-item-link href=/tidyomicsBlog/categories/>Categories</a></li><li class=menu-item><a class=menu-item-link href=/tidyomicsBlog/about/>About</a></li></ul></nav></header><main id=main class=main><div class=content-wrapper><div id=content class=content><article class=post><header class=post-header><h1 class=post-title>Tidy-transcriptomics manifesto</h1><div class=post-meta><span class=post-time>2021-07-07</span><div class=post-category><a href=/tidyomicsBlog/categories/all-platforms/>all platforms</a></div></div></header><div class=post-content><pre><code>## ‚îÄ‚îÄ Attaching packages ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ tidyverse 1.3.1 ‚îÄ‚îÄ</code></pre><pre><code>## ‚úì ggplot2 3.3.5     ‚úì purrr   0.3.4
## ‚úì tibble  3.1.2     ‚úì dplyr   1.0.7
## ‚úì tidyr   1.1.3     ‚úì stringr 1.4.0
## ‚úì readr   1.4.0     ‚úì forcats 0.5.1</code></pre><pre><code>## ‚îÄ‚îÄ Conflicts ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ tidyverse_conflicts() ‚îÄ‚îÄ
## x dplyr::filter() masks stats::filter()
## x dplyr::lag()    masks stats::lag()</code></pre><p><img src=images/logo.png alt=tidyomics/tidyomicsBlog_logo width=100px height=100px></p><p>In this first post, we want to introduce the concept of tidy transcriptomics and its principles is moved by.</p><div id=what-is-tidy-transcriptomics class="section level1"><h1>What is tidy transcriptomics</h1><p>Tidy transcriptomics is a specific approach to transcriptomic data analysis in R, that uses the ‚Äútidy‚Äù principles proposed by <a href=https://joss.theoj.org/papers/10.21105/joss.01686.pdf>Wickham et al.</a>. Tidy transcriptomics introduces both a tidy data representation and manipulation (for single-cell and bulk) with the packages <code>tidySummarizedExperiment</code>, <code>tidySingleCellExperiment</code>, and <code>tidyseurat</code> and a tidy analysis workflow (for bulk data) with the package <code>tidybulk</code>.</p></div><div id=the-tidy-transcriptomics-manifesto class="section level1"><h1>The tidy transcriptomics manifesto</h1><p>This manifesto lays down the principles of tidy transcriptomics that <code>tidybulk</code> and <code>tidySummarizedExperiment</code>, <code>tidySingleCellExperiment</code>, and <code>tidyseurat</code> are based on. These principles are in line with <a href=https://cran.r-project.org/web/packages/tidyverse/vignettes/manifesto.html>the tidy tools manifesto</a>.</p><ul><li><strong>Use easy-to-understand, verbose, jargon- and acronym-free vocabulary.</strong> If the English dictionary is not enough to understand the underlying meaning of a function or a variable name, it is a bad sign. _<a href=https://iubmb.onlinelibrary.wiley.com/doi/10.1002/bmb.20922_ class=uri>https://iubmb.onlinelibrary.wiley.com/doi/10.1002/bmb.20922_</a>; and the cost of saving a few characters may be bigger than we think.</li><li><strong>Present/visualise information in its raw form when possible.</strong> Modern visualisation tools (e.g.¬†ggplot2) allow the use of custom scales to visualise data. For example, for visualising p-values apply the <code>log10_reverse</code> scale instead of transforming the p-values in their negative-log form; for visualising (raw, scaled and/or adjusted) transcript abundance (in the form of read counts) use the <code>log1p</code> scale instead of transforming the data in its log (or log count-per-million) form.</li><li><strong>Avoid the creation of temporary variables when possible.</strong> When working interactively, creating variables to store data is a bug-prone process, especially if multiple assignments are done on the same variable through the workflow. The main utility of variables should be to store data that is used more than once. Tidyverse allows complex operations to be combined in a simple way, reducing the need to create temporary variables.</li></ul></div><div id=integration-map class="section level1"><h1>Integration map</h1><p>All packages are under active development, and are in a <a href=https://lifecycle.r-lib.org/articles/stages.html>maturing</a> lifecycle.</p><p><img src=images/roadmap_integration.png alt=roadmap width=90%></p><p>There are two parts in the tidy transcriptomics ecosystem: data and analysis framework. So far, for bulk RNA sequencing both data (tidySummarizedExperiment) and analysis (tidybulk) frameworks are available. In contrast,for single cell only data frameworks are available (tidySingleCellExperiment and tidyseurat).</p><div id=what-data-frameworks-are-and-what-they-are-not class="section level2"><h2>What data frameworks are and what they are not</h2><p>Data frameworks are not data containers. Data frameworks are data-abstraction that display and manipulate your existing containers (i.e.¬†<code>SummarizedExperiment</code>, <code>SingleCellExperiment</code> and <code>Seurat</code> object) in a tidy manner. Therefore there is not such a thing as <code>tidy*</code> object. This has the advantage of allowing you to use tidyverse on transcriptomics data without compromising your existing pipelines. That is, if a method works for <code>SummarizedExperiment</code> it works for its tidy representation.</p><p>Therefore, the question ‚Äúcan we go from <code>tidySummarizedExperiment</code> to <code>SummarizedExperiment</code> and viceversa‚Äù is not relevant, as we never leave <code>SummarizedExperiment</code> in the first place.</p></div><div id=giving-a-consistent-interface-despite-different-data-containers class="section level2"><h2>Giving a consistent interface despite different data containers</h2><p>With tidy transcriptomics, we differentiate the data container from the user interface. As an analogy, if we want to see picture of cats, we don‚Äôt care that a Unix and Windows machines store information in the hard drive differently. Similarly, if we want to display, manipulate and visualise transcriptomic data, we might not care how the data is stored.</p></div><div id=which-analysis-framework-can-interface-with-which-data-framework class="section level2"><h2>Which analysis framework can interface with which data framework</h2><p><strong>Bulk RNA sequencing data</strong></p><ul><li><p><code>SummarizedExperiment</code> can interface with <code>Bioconductor</code> and <code>tidybulk</code></p></li><li><p><code>tibble</code> can interface with tidyverse and <code>tidybulk</code></p></li><li><p><code>tidySummarizedExperiment</code> can interface with all three: <code>Bioconductor</code>, tidyverse and <code>tidybulk</code></p></li></ul><p><strong>Single-cell RNA sequencing data</strong></p><ul><li><p><code>SingleCellExperiment</code> can interface with <code>Bioconductor</code></p></li><li><p><code>tidySingleCellExperiment</code> can interface with <code>Bioconductor</code> and <code>tidyverse</code></p></li><li><p><code>Seurat</code> object can interface with <code>Seurat</code></p></li><li><p><code>tidyseurat</code> object can interface with <code>Seurat</code> and <code>tidyverse</code></p></li></ul></div></div><div id=differences-between-tidy-and-base-transcriptomics-sinthax class="section level1"><h1>Differences between tidy and base transcriptomics sinthax</h1><p>Here we provide few examples of the differences in programming transcriptomics analyses with tidy transcriptomics and the base R alternative. An important aspect is that tidy transcriptomics is complementary with standard workflows. Everything that works with <code>SummarizedExperiment</code>, <code>SingleCellExperiment</code> and <code>Seurat</code> works with their tidy representations.</p><div id=tidybulk-and-tidysummarizedexperiment class="section level2"><h2>Tidybulk and tidySummarizedExperiment</h2><p>This example is taken from the workshop (BioC2021)[<a href=https://stemangiola.github.io/bioc2021_tidyomics/tidyomicsBlog/articles/tidyomics/tidyomicsBlog.html class=uri>https://stemangiola.github.io/bioc2021_tidyomics/tidyomicsBlog/articles/tidyomics/tidyomicsBlog.html</a>].</p><p>As tidybulk smoothly integrates with ggplot2 and other tidyverse packages it can save on typing and make plots easier to generate. Compare the code for creating density plots with tidybulk versus standard base R below (standard code adapted from (Law et al.¬†2016)).</p><p><strong>tidybulk</strong></p><pre class=r><code># tidybulk
airway %&gt;%
  keep_abundant(factor_of_interest = dex) %&gt;%
  scale_abundance() %&gt;%
  pivot_longer(cols = c(&quot;counts&quot;, &quot;counts_scaled&quot;), names_to = &quot;source&quot;, values_to = &quot;abundance&quot;) %&gt;%
  ggplot(aes(x = abundance + 1, color = sample)) +
  geom_density() +
  facet_wrap(~source) +
  scale_x_log10() +
  custom_theme</code></pre><p><strong>base R using edgeR</strong></p><pre class=r><code># Example code, no need to run
# Prepare data set
library(edgeR)
dgList &lt;- SE2DGEList(airway)
group &lt;- factor(dgList$samples$dex)
keep.exprs &lt;- filterByExpr(dgList, group = group)
dgList &lt;- dgList[keep.exprs, , keep.lib.sizes = FALSE]
nsamples &lt;- ncol(dgList)
logcounts &lt;- log2(dgList$counts)
# Setup graphics
col &lt;- RColorBrewer::brewer.pal(nsamples, &quot;Paired&quot;)
par(mfrow = c(1, 2))
# Plot raw counts
plot(density(logcounts[, 1]), col = col[1], lwd = 2, ylim = c(0, 0.26), las = 2, main = &quot;&quot;, xlab = &quot;&quot;)
title(main = &quot;Counts&quot;)
for (i in 2:nsamples) {
  den &lt;- density(logcounts[, i])
  lines(den$x, den$y, col = col[i], lwd = 2)
}
legend(&quot;topright&quot;, legend = dgList$samples$Run, text.col = col, bty = &quot;n&quot;)
# Plot scaled counts
dgList_norm &lt;- calcNormFactors(dgList)
lcpm_n &lt;- cpm(dgList_norm, log = TRUE)
plot(density(lcpm_n[, 1]), col = col[1], lwd = 2, ylim = c(0, 0.26), las = 2, main = &quot;&quot;, xlab = &quot;&quot;)
title(&quot;Counts scaled&quot;)
for (i in 2:nsamples) {
  den &lt;- density(lcpm_n[, i])
  lines(den$x, den$y, col = col[i], lwd = 2)
}
legend(&quot;topright&quot;, legend = dgList_norm$samples$Run, text.col = col, bty = &quot;n&quot;)</code></pre></div><div id=singlecellexperiment-and-tidysinglecellexperiment class="section level2"><h2>SingleCellExperiment and tidySingleCellExperiment</h2><p>These examples are taken from the article (Interfacing Seurat with the R tidy universe)[<a href=https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btab404/6283576 class=uri>https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btab404/6283576</a>].</p><p>Calculating a transcriptional signature.</p><p><strong>Base</strong></p><pre class=r><code># Load data
pbmc_SCE = tidySingleCellExperiment::pbmc_small

signature_score_1 =
  pbmc_SCE[c(&quot;CD79A&quot;, &quot;CD79B&quot;),] %&gt;%
  logcounts() %&gt;%
  colSums() %&gt;%
  scales::rescale(to = c(0,1))

signature_score_2 = 
  pbmc_SCE[c(&quot;CD3D&quot;, &quot;CCR7&quot;),] %&gt;%
  logcounts() %&gt;%
  colSums() %&gt;%
  scales::rescale(to = c(0,1))

signature_score_1 - signature_score_2</code></pre><pre><code>## ATGCCAGAACGACT CATGGCCTGTGCAT GAACCTGATGAACC TGACTGGATTCTCA AGTCAGACTGCACA 
##     -0.1282987     -1.0000000     -0.4892647     -0.4769175     -0.8142420 
## TCTGATACACGTGT TGGTATCTAAACAG GCAGCTCTGTTTCT GATATAACACGCAT AATGTTGACAGTCA 
##     -0.5065754     -0.8943038     -0.8433443     -0.4750597     -0.4399134 
## AGGTCATGAGTGTC AGAGATGATCTCGC GGGTAACTCTAGTG CATGAGACACGGGA TACGCCACTCCGAA 
##      0.4399898      0.8313740      0.8367149      0.9404373      0.8785365 
## CTAAACCTGTGCAT GTAAGCACTCATTC TTGGTACTGAATCC CATCATACGGAGCA TACATCACGCTAAC 
##      0.8647225      0.7631773      0.8315032      0.9127912      0.6386232 
## TTACCATGAATCGC ATAGGAGAAACAGA GCGCACGACTTTAC ACTCGCACGAAAGT ATTACCTGCCTTAT 
##      0.0000000      0.0000000      0.2946328      0.0000000     -0.2487348 
## CCCAACTGCAATCG AAATTCGAATCACG CCATCCGATTCGCC TCCACTCTGAGCTT CATCAGGATGCACA 
##      0.0000000      0.0000000      0.2908813      0.0000000      0.2572103 
## CTAAACCTCTGACA GATAGAGAAGGGTG CTAACGGAACCGAT AGATATACCCGTAA TACTCTGAATCGAC 
##      0.2839127     -0.5116071      0.3556147      0.0000000      0.0000000 
## GCGCATCTTGCTCC GTTGACGATATCGG ACAGGTACTGGTGT GGCATATGCTTATC CATTACACCAACTG 
##      0.0000000     -0.3058682      0.0000000     -0.3499484      0.0000000 
## TAGGGACTGAACTC GCTCCATGAGAAGT TACAATGATGCTAG CTTCATGACCGAAT CTGCCAACAGGAGC 
##      0.0000000     -0.3969658     -0.4491779      0.0000000     -0.4252278 
## TTGCATTGAGCTAC AAGCAAGAGCTTAG CGGCACGAACTCAG GGTGGAGATTACTC GGCCGATGTACTCT 
##     -0.5803536     -0.3499484     -0.4602168     -0.4537220     -0.4789804 
## CGTAGCCTGTATGC TGAGCTGAATGCTG CCTATAACGAGACG ATAAGTTGGTACGT AAGCGACTTTGACG 
##      0.2535563      0.0000000      0.3265729      0.3521070      0.2917172 
## ACCAGTGAATACCG ATTGCACTTGCTTT CTAGGTGATGGTTG GCACTAGACCTTTA CATGCGCTAGTCAC 
##      0.2962249      0.0000000      0.0000000      0.3533870      0.0000000 
## TTGAGGACTACGCA ATACCACTCTAAGC CATATAGACTAAGC TTTAGCTGTACTCT GACATTCTCCACCT 
##      0.0000000      0.0000000      0.0000000      0.7333251      0.1920881 
## ACGTGATGCCATGA ATTGTAGATTCCCG GATAGAGATCACGA AATGCGTGGACGGA GCGTAAACACGGTT 
##      0.0000000      0.0000000      0.0000000      0.2500814     -0.2120188 
## ATTCAGCTCATTGG GGCATATGGGGAGT ATCATCTGACACCA GTCATACTTCGCCT TTACGTACGTTCAG 
##      0.0000000      0.0000000      0.0000000      0.0000000      0.0000000 
## GAGTTGTGGTAGCT GACGCTCTCTCTCG AGTCTTACTTCGGA GGAACACTTCAGAC CTTGATTGATCTTC 
##      0.0000000      0.0000000      0.0000000      0.0000000      0.0000000</code></pre><p><strong>tidySingleCellExperiment</strong></p><pre class=r><code># Load data
pbmc_SCE = tidySingleCellExperiment::pbmc_small

pbmc_SCE %&gt;%
  join_features(c(&quot;CD79A&quot;, &quot;CD79B&quot;, &quot;CD3D&quot;, &quot;CCR7&quot;), assay = &quot;logcounts&quot;, shape = &quot;wide&quot;) %&gt;%
  mutate(signature_score = 
           scales::rescale(CD79A + CD79B, to = c(0,1)) - 
           scales::rescale(CD3D + CCR7, to = c(0,1))
  ) %&gt;%
  select(signature_score, everything())</code></pre><pre><code>## # A SingleCellExperiment-tibble abstraction: 80 x 22
## [90m# Features=230 | Assays=counts, logcounts[39m
##    cell  signature_score orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8
##    &lt;chr&gt;           &lt;dbl&gt; &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;          
##  1 ATGC‚Ä¶          -0.128 SeuratPro‚Ä¶         70           47 0              
##  2 CATG‚Ä¶          -1     SeuratPro‚Ä¶         85           52 0              
##  3 GAAC‚Ä¶          -0.489 SeuratPro‚Ä¶         87           50 1              
##  4 TGAC‚Ä¶          -0.477 SeuratPro‚Ä¶        127           56 0              
##  5 AGTC‚Ä¶          -0.814 SeuratPro‚Ä¶        173           53 0              
##  6 TCTG‚Ä¶          -0.507 SeuratPro‚Ä¶         70           48 0              
##  7 TGGT‚Ä¶          -0.894 SeuratPro‚Ä¶         64           36 0              
##  8 GCAG‚Ä¶          -0.843 SeuratPro‚Ä¶         72           45 0              
##  9 GATA‚Ä¶          -0.475 SeuratPro‚Ä¶         52           36 0              
## 10 AATG‚Ä¶          -0.440 SeuratPro‚Ä¶        100           41 0              
## # ‚Ä¶ with 70 more rows, and 16 more variables: letter.idents &lt;fct&gt;,
## #   groups &lt;chr&gt;, RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;, ident &lt;fct&gt;, CD79A &lt;dbl&gt;,
## #   CD79B &lt;dbl&gt;, CD3D &lt;dbl&gt;, CCR7 &lt;dbl&gt;, PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;, PC_3 &lt;dbl&gt;,
## #   PC_4 &lt;dbl&gt;, PC_5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;</code></pre><p>Cell subsampling for mulisample balanced UMAP representation</p><p><strong>Base</strong></p><pre class=r><code>splits = 
  colnames(pbmc_SCE) %&gt;%
  split(pbmc_SCE$file  )

min_size = splits %&gt;%
  sapply(length) %&gt;% min()

cell_subset = splits %&gt;%
  lapply(function(x) sample(x, min_size)) %&gt;%
  unlist()

pbmc_SCE[, cell_subset]</code></pre><pre><code>## # A SingleCellExperiment-tibble abstraction: 72 x 17
## [90m# Features=230 | Assays=counts, logcounts[39m
##    cell  orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents groups
##    &lt;chr&gt; &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           &lt;fct&gt;         &lt;chr&gt; 
##  1 ATTA‚Ä¶ SeuratPro‚Ä¶        463           77 1               B             g1    
##  2 CTAA‚Ä¶ SeuratPro‚Ä¶        246           59 0               A             g1    
##  3 TAGG‚Ä¶ SeuratPro‚Ä¶        156           60 0               A             g1    
##  4 ATAG‚Ä¶ SeuratPro‚Ä¶        406           74 1               B             g1    
##  5 AATG‚Ä¶ SeuratPro‚Ä¶        100           41 0               A             g1    
##  6 GGCA‚Ä¶ SeuratPro‚Ä¶        172           29 0               A             g1    
##  7 TACT‚Ä¶ SeuratPro‚Ä¶        156           48 0               A             g1    
##  8 CTTG‚Ä¶ SeuratPro‚Ä¶        233           76 1               B             g1    
##  9 GATA‚Ä¶ SeuratPro‚Ä¶        328           72 1               B             g1    
## 10 ACAG‚Ä¶ SeuratPro‚Ä¶        151           59 0               A             g1    
## # ‚Ä¶ with 62 more rows, and 10 more variables: RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;,
## #   ident &lt;fct&gt;, PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;, PC_3 &lt;dbl&gt;, PC_4 &lt;dbl&gt;, PC_5 &lt;dbl&gt;,
## #   tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;</code></pre><p><strong>tidySingleCellExperiment</strong></p><pre class=r><code>pbmc_SCE %&gt;%
  add_count(file  , name = &quot;tot_cells&quot;) %&gt;%
  mutate(min_cells = min(tot_cells)) %&gt;%
  group_by(file  ) %&gt;%
  sample_n(min_cells)</code></pre><pre><code>## Warning: The `.drop` argument of `add_count()` is deprecated as of dplyr 1.0.0.</code></pre><pre><code>## tidySingleCellExperiment says: A data frame is returned for independent data analysis.</code></pre><pre><code>## # A tibble: 72 x 33
## # Groups:   file [2]
##    cell  orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents groups
##    &lt;chr&gt; &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           &lt;fct&gt;         &lt;chr&gt; 
##  1 TTGA‚Ä¶ SeuratPro‚Ä¶        787           88 0               A             g1    
##  2 CTAA‚Ä¶ SeuratPro‚Ä¶        189           53 0               A             g1    
##  3 TGGT‚Ä¶ SeuratPro‚Ä¶         64           36 0               A             g1    
##  4 ATTA‚Ä¶ SeuratPro‚Ä¶        463           77 1               B             g1    
##  5 GGCA‚Ä¶ SeuratPro‚Ä¶        172           29 0               A             g1    
##  6 TTGG‚Ä¶ SeuratPro‚Ä¶        135           45 0               A             g1    
##  7 GCAG‚Ä¶ SeuratPro‚Ä¶         72           45 0               A             g1    
##  8 CTGC‚Ä¶ SeuratPro‚Ä¶        146           47 0               A             g1    
##  9 CTTG‚Ä¶ SeuratPro‚Ä¶        233           76 1               B             g1    
## 10 ACCA‚Ä¶ SeuratPro‚Ä¶        417           75 0               A             g1    
## # ‚Ä¶ with 62 more rows, and 26 more variables: RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;,
## #   ident &lt;fct&gt;, tot_cells &lt;int&gt;, min_cells &lt;int&gt;, PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;,
## #   PC_3 &lt;dbl&gt;, PC_4 &lt;dbl&gt;, PC_5 &lt;dbl&gt;, PC_6 &lt;dbl&gt;, PC_7 &lt;dbl&gt;, PC_8 &lt;dbl&gt;,
## #   PC_9 &lt;dbl&gt;, PC_10 &lt;dbl&gt;, PC_11 &lt;dbl&gt;, PC_12 &lt;dbl&gt;, PC_13 &lt;dbl&gt;,
## #   PC_14 &lt;dbl&gt;, PC_15 &lt;dbl&gt;, PC_16 &lt;dbl&gt;, PC_17 &lt;dbl&gt;, PC_18 &lt;dbl&gt;,
## #   PC_19 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;</code></pre></div><div id=tidyseurat class="section level2"><h2>tidyseurat</h2><p>Calculating a transcriptional signature.</p><p><strong>Base</strong></p><pre class=r><code>library(Seurat)</code></pre><pre><code>## Registered S3 method overwritten by &#39;spatstat.geom&#39;:
##   method     from
##   print.boxx cli</code></pre><pre><code>## Attaching SeuratObject</code></pre><pre><code>## 
## Attaching package: &#39;Seurat&#39;</code></pre><pre><code>## The following object is masked from &#39;package:SummarizedExperiment&#39;:
## 
##     Assays</code></pre><pre class=r><code># Load data
pbmc_Seurat = tidySingleCellExperiment::pbmc_small %&gt;% as.Seurat() %&gt;% NormalizeData()

signature_score_1 =
  pbmc_Seurat[c(&quot;CD79A&quot;, &quot;CD79B&quot;),] %&gt;%
  Seurat::GetAssayData() %&gt;%
  colSums() %&gt;%
  scales::rescale(to = c(0,1))

signature_score_2 = 
  pbmc_Seurat[c(&quot;CD3D&quot;, &quot;CCR7&quot;),] %&gt;%
  Seurat::GetAssayData() %&gt;%
  colSums() %&gt;%
  scales::rescale(to = c(0,1))

signature_score_1 - signature_score_2</code></pre><pre><code>## ATGCCAGAACGACT CATGGCCTGTGCAT GAACCTGATGAACC TGACTGGATTCTCA AGTCAGACTGCACA 
##     -0.1282987     -1.0000000     -0.4892647     -0.4769175     -0.8142420 
## TCTGATACACGTGT TGGTATCTAAACAG GCAGCTCTGTTTCT GATATAACACGCAT AATGTTGACAGTCA 
##     -0.5065754     -0.8943038     -0.8433443     -0.4750597     -0.4399134 
## AGGTCATGAGTGTC AGAGATGATCTCGC GGGTAACTCTAGTG CATGAGACACGGGA TACGCCACTCCGAA 
##      0.4399898      0.8313740      0.8367149      0.9404373      0.8785365 
## CTAAACCTGTGCAT GTAAGCACTCATTC TTGGTACTGAATCC CATCATACGGAGCA TACATCACGCTAAC 
##      0.8647225      0.7631773      0.8315032      0.9127912      0.6386232 
## TTACCATGAATCGC ATAGGAGAAACAGA GCGCACGACTTTAC ACTCGCACGAAAGT ATTACCTGCCTTAT 
##      0.0000000      0.0000000      0.2946328      0.0000000     -0.2487348 
## CCCAACTGCAATCG AAATTCGAATCACG CCATCCGATTCGCC TCCACTCTGAGCTT CATCAGGATGCACA 
##      0.0000000      0.0000000      0.2908813      0.0000000      0.2572103 
## CTAAACCTCTGACA GATAGAGAAGGGTG CTAACGGAACCGAT AGATATACCCGTAA TACTCTGAATCGAC 
##      0.2839127     -0.5116071      0.3556147      0.0000000      0.0000000 
## GCGCATCTTGCTCC GTTGACGATATCGG ACAGGTACTGGTGT GGCATATGCTTATC CATTACACCAACTG 
##      0.0000000     -0.3058682      0.0000000     -0.3499484      0.0000000 
## TAGGGACTGAACTC GCTCCATGAGAAGT TACAATGATGCTAG CTTCATGACCGAAT CTGCCAACAGGAGC 
##      0.0000000     -0.3969658     -0.4491779      0.0000000     -0.4252278 
## TTGCATTGAGCTAC AAGCAAGAGCTTAG CGGCACGAACTCAG GGTGGAGATTACTC GGCCGATGTACTCT 
##     -0.5803536     -0.3499484     -0.4602168     -0.4537220     -0.4789804 
## CGTAGCCTGTATGC TGAGCTGAATGCTG CCTATAACGAGACG ATAAGTTGGTACGT AAGCGACTTTGACG 
##      0.2535563      0.0000000      0.3265729      0.3521070      0.2917172 
## ACCAGTGAATACCG ATTGCACTTGCTTT CTAGGTGATGGTTG GCACTAGACCTTTA CATGCGCTAGTCAC 
##      0.2962249      0.0000000      0.0000000      0.3533870      0.0000000 
## TTGAGGACTACGCA ATACCACTCTAAGC CATATAGACTAAGC TTTAGCTGTACTCT GACATTCTCCACCT 
##      0.0000000      0.0000000      0.0000000      0.7333251      0.1920881 
## ACGTGATGCCATGA ATTGTAGATTCCCG GATAGAGATCACGA AATGCGTGGACGGA GCGTAAACACGGTT 
##      0.0000000      0.0000000      0.0000000      0.2500814     -0.2120188 
## ATTCAGCTCATTGG GGCATATGGGGAGT ATCATCTGACACCA GTCATACTTCGCCT TTACGTACGTTCAG 
##      0.0000000      0.0000000      0.0000000      0.0000000      0.0000000 
## GAGTTGTGGTAGCT GACGCTCTCTCTCG AGTCTTACTTCGGA GGAACACTTCAGAC CTTGATTGATCTTC 
##      0.0000000      0.0000000      0.0000000      0.0000000      0.0000000</code></pre><p><strong>tidyseurat</strong></p><p>It is the same as for <code>tidySingleCellExperiment</code></p><pre class=r><code>library(tidyseurat)</code></pre><pre><code>## ========================================
## tidyseurat version 0.2.3
## If you use TIDYSEURAT in published research, please cite:
## 
## Mangiola et al. Interfacing Seurat with the R tidy universe. Bioinformatics 2021.
## 
## This message can be suppressed by:
##   suppressPackageStartupMessages(library(tidyseurat))
##   
## To restore the Seurat default display use options(&quot;restore_Seurat_show&quot; = TRUE) 
## ========================================</code></pre><pre><code>## 
## Attaching package: &#39;tidyseurat&#39;</code></pre><pre><code>## The following object is masked from &#39;package:IRanges&#39;:
## 
##     slice</code></pre><pre><code>## The following object is masked from &#39;package:S4Vectors&#39;:
## 
##     rename</code></pre><pre><code>## The following object is masked from &#39;package:matrixStats&#39;:
## 
##     count</code></pre><pre><code>## The following objects are masked from &#39;package:dplyr&#39;:
## 
##     add_count, bind_cols, bind_rows, count</code></pre><pre><code>## The following object is masked from &#39;package:stats&#39;:
## 
##     filter</code></pre><pre class=r><code>pbmc_Seurat %&gt;%
  join_features(c(&quot;CD79A&quot;, &quot;CD79B&quot;, &quot;CD3D&quot;, &quot;CCR7&quot;), shape = &quot;wide&quot;) %&gt;%
  mutate(signature_score = 
           scales::rescale(CD79A + CD79B, to = c(0,1)) - 
           scales::rescale(CD3D + CCR7, to = c(0,1))
  ) %&gt;%
  select(signature_score, everything())</code></pre><pre><code>## # A Seurat-tibble abstraction: 80 x 24
## [90m# Features=230 | Active assay=originalexp | Assays=originalexp[39m
##    cell  signature_score orig.ident nCount_original‚Ä¶ nFeature_origin‚Ä¶ nCount_RNA
##    &lt;chr&gt;           &lt;dbl&gt; &lt;fct&gt;                 &lt;dbl&gt;            &lt;int&gt;      &lt;dbl&gt;
##  1 ATGC‚Ä¶          -0.128 SeuratPro‚Ä¶               70               47         70
##  2 CATG‚Ä¶          -1     SeuratPro‚Ä¶               85               52         85
##  3 GAAC‚Ä¶          -0.489 SeuratPro‚Ä¶               87               50         87
##  4 TGAC‚Ä¶          -0.477 SeuratPro‚Ä¶              127               56        127
##  5 AGTC‚Ä¶          -0.814 SeuratPro‚Ä¶              173               53        173
##  6 TCTG‚Ä¶          -0.507 SeuratPro‚Ä¶               70               48         70
##  7 TGGT‚Ä¶          -0.894 SeuratPro‚Ä¶               64               36         64
##  8 GCAG‚Ä¶          -0.843 SeuratPro‚Ä¶               72               45         72
##  9 GATA‚Ä¶          -0.475 SeuratPro‚Ä¶               52               36         52
## 10 AATG‚Ä¶          -0.440 SeuratPro‚Ä¶              100               41        100
## # ‚Ä¶ with 70 more rows, and 18 more variables: nFeature_RNA &lt;int&gt;,
## #   RNA_snn_res.0.8 &lt;fct&gt;, letter.idents &lt;fct&gt;, groups &lt;chr&gt;,
## #   RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;, ident &lt;fct&gt;, CD79A &lt;dbl&gt;, CD79B &lt;dbl&gt;,
## #   CD3D &lt;dbl&gt;, CCR7 &lt;dbl&gt;, PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;, PC_3 &lt;dbl&gt;, PC_4 &lt;dbl&gt;,
## #   PC_5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;</code></pre><p>Cell subsampling for mulisample balanced UMAP representation</p><p><strong>Base</strong></p><pre class=r><code>splits = 
  colnames(pbmc_Seurat) %&gt;%
  split(pbmc_Seurat$file  )

min_size = splits %&gt;%
  sapply(length) %&gt;% min()

cell_subset = splits %&gt;%
  lapply(function(x) sample(x, min_size)) %&gt;%
  unlist()

pbmc_Seurat[, cell_subset]</code></pre><pre><code>## # A Seurat-tibble abstraction: 72 x 19
## [90m# Features=230 | Active assay=originalexp | Assays=originalexp[39m
##    cell  orig.ident nCount_original‚Ä¶ nFeature_origin‚Ä¶ nCount_RNA nFeature_RNA
##    &lt;chr&gt; &lt;fct&gt;                 &lt;dbl&gt;            &lt;int&gt;      &lt;dbl&gt;        &lt;int&gt;
##  1 GCAG‚Ä¶ SeuratPro‚Ä¶               72               45         72           45
##  2 GCGT‚Ä¶ SeuratPro‚Ä¶              754               83        754           83
##  3 CATG‚Ä¶ SeuratPro‚Ä¶               85               52         85           52
##  4 CATA‚Ä¶ SeuratPro‚Ä¶              286               68        286           68
##  5 CTGC‚Ä¶ SeuratPro‚Ä¶              146               47        146           47
##  6 CATC‚Ä¶ SeuratPro‚Ä¶              353               80        353           80
##  7 TACT‚Ä¶ SeuratPro‚Ä¶              156               48        156           48
##  8 ATAC‚Ä¶ SeuratPro‚Ä¶              612               69        612           69
##  9 GGCA‚Ä¶ SeuratPro‚Ä¶              126               53        126           53
## 10 GACA‚Ä¶ SeuratPro‚Ä¶              872               96        872           96
## # ‚Ä¶ with 62 more rows, and 13 more variables: RNA_snn_res.0.8 &lt;fct&gt;,
## #   letter.idents &lt;fct&gt;, groups &lt;chr&gt;, RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;,
## #   ident &lt;fct&gt;, PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;, PC_3 &lt;dbl&gt;, PC_4 &lt;dbl&gt;, PC_5 &lt;dbl&gt;,
## #   tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;</code></pre><p><strong>tidyseurat</strong></p><p>It is the same as for <code>tidySingleCellExperiment</code></p><pre class=r><code>pbmc_Seurat %&gt;%
  add_count(file  , name = &quot;tot_cells&quot;) %&gt;%
  mutate(min_cells = min(tot_cells)) %&gt;%
  group_by(file  ) %&gt;%
  sample_n(min_cells)</code></pre><pre><code>## tidyseurat says: A data frame is returned for independent data analysis.</code></pre><pre><code>## # A tibble: 72 x 35
## # Groups:   file [2]
##    cell    orig.ident  nCount_original‚Ä¶ nFeature_origin‚Ä¶ nCount_RNA nFeature_RNA
##    &lt;chr&gt;   &lt;fct&gt;                  &lt;dbl&gt;            &lt;int&gt;      &lt;dbl&gt;        &lt;int&gt;
##  1 GCAGCT‚Ä¶ SeuratProj‚Ä¶               72               45         72           45
##  2 GCGTAA‚Ä¶ SeuratProj‚Ä¶              754               83        754           83
##  3 CATGGC‚Ä¶ SeuratProj‚Ä¶               85               52         85           52
##  4 CATATA‚Ä¶ SeuratProj‚Ä¶              286               68        286           68
##  5 CTGCCA‚Ä¶ SeuratProj‚Ä¶              146               47        146           47
##  6 CATCAG‚Ä¶ SeuratProj‚Ä¶              353               80        353           80
##  7 TACTCT‚Ä¶ SeuratProj‚Ä¶              156               48        156           48
##  8 ATACCA‚Ä¶ SeuratProj‚Ä¶              612               69        612           69
##  9 GGCATA‚Ä¶ SeuratProj‚Ä¶              126               53        126           53
## 10 GACATT‚Ä¶ SeuratProj‚Ä¶              872               96        872           96
## # ‚Ä¶ with 62 more rows, and 29 more variables: RNA_snn_res.0.8 &lt;fct&gt;,
## #   letter.idents &lt;fct&gt;, groups &lt;chr&gt;, RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;,
## #   ident &lt;fct&gt;, tot_cells &lt;int&gt;, min_cells &lt;int&gt;, PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;,
## #   PC_3 &lt;dbl&gt;, PC_4 &lt;dbl&gt;, PC_5 &lt;dbl&gt;, PC_6 &lt;dbl&gt;, PC_7 &lt;dbl&gt;, PC_8 &lt;dbl&gt;,
## #   PC_9 &lt;dbl&gt;, PC_10 &lt;dbl&gt;, PC_11 &lt;dbl&gt;, PC_12 &lt;dbl&gt;, PC_13 &lt;dbl&gt;,
## #   PC_14 &lt;dbl&gt;, PC_15 &lt;dbl&gt;, PC_16 &lt;dbl&gt;, PC_17 &lt;dbl&gt;, PC_18 &lt;dbl&gt;,
## #   PC_19 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;</code></pre></div></div><div id=learning-material class="section level1"><h1>Learning material</h1><p>We delivered a series of workshops on tidy transcriptomics, which track the progress of the ecosystem.</p><p><a href=https://stemangiola.github.io/bioc2021_tidyomics/tidyomicsBlog/index.html>Bioc 2021</a></p><p><a href=https://tidyomics/tidyomicsBlog-workshops.github.io/ismb2021_tidyomics/tidyomicsBlog/index.html>ISMB/ECCB 2021</a></p><p><a href=https://stemangiola.github.io/rladiestunis2021_tidyomics/tidyomicsBlog/index.html>R-Ladies Tunis 2021</a></p><p><a href=https://stemangiola.github.io/bioceurope2020_tidyomics/tidyomicsBlog/index.html>Euro Bioc 2020</a></p><p><a href=https://stemangiola.github.io/ABACBS2020_tidyomics/tidyomicsBlog/index.html>ABACBS 2020</a></p><p><a href=https://stemangiola.github.io/biocasia2020_tidyomics/tidyomicsBlog/index.html>Bioc Asia 2020</a></p><p><a href=https://stemangiola.github.io/rpharma2020_tidyomics/tidyomicsBlog/index.html>R/Pharma 2020</a></p><p><a href=https://stemangiola.github.io/bioc_2020_tidyomics/tidyomicsBlog/index.html>Bioc 2020</a></p></div><div id=how-to-cite-tidy-transcriptomics class="section level1"><h1>How to cite tidy transcriptomics</h1><div id=tidybulk class="section level3"><h3>tidybulk</h3><p>Mangiola, S., Molania, R., Dong, R. et al.¬†tidybulk: an R tidy framework for modular transcriptomic data analysis. Genome Biol 22, 42 (2021). <a href=https://doi.org/10.1186/s13059-020-02233-7 class=uri>https://doi.org/10.1186/s13059-020-02233-7</a></p></div><div id=tidyseurat-1 class="section level3"><h3>tidyseurat</h3><p>Stefano Mangiola, Maria A Doyle, Anthony T Papenfuss, Interfacing Seurat with the R tidy universe, Bioinformatics, 2021;, btab404, <a href=https://doi.org/10.1093/bioinformatics/btab404 class=uri>https://doi.org/10.1093/bioinformatics/btab404</a></p></div><div id=tidysummarizedexperiment-and-tidysinglecellexeperiment class="section level3"><h3>tidySummarizedExperiment and tidySingleCellEXeperiment</h3><p>You can use <code>tidyseurat</code> citation as introduces the concepts of data abstraction.</p></div></div></div><footer class=post-footer><div class=post-tags><a href=/tidyomicsBlog/tags/tidyomics/tidyomicsblog/>tidyomics/tidyomicsBlog</a>
<a href=/tidyomicsBlog/tags/manifesto/>manifesto</a>
<a href=/tidyomicsBlog/tags/transcriptomics/>transcriptomics</a>
<a href=/tidyomicsBlog/tags/tidyverse/>tidyverse</a></div><nav class=post-nav><a class=prev href=/tidyomicsBlog/post/2023-12-07-tidy-sccomp/><i class="iconfont icon-left"></i>
<span class="prev-text nav-default">The new tidy sccomp interface</span>
<span class="prev-text nav-mobile">Prev</span></a></nav></footer></article></div></div></main><footer id=footer class=footer><div class=social-links><a href=mailto:mangiolastefano@gmail.com class="iconfont icon-email" title=email></a>
<a href=https://github.com/stemangiola class="iconfont icon-github" title=github></a>
<a href=https://twitter.com/steman_research class="iconfont icon-twitter" title=twitter></a>
<a href=https://tidyomics.github.io/tidyomicsBlog/index.xml type=application/rss+xml class="iconfont icon-rss" title=rss></a></div><div class=copyright><span class=power-by>Powered by <a class=hexo-link href=https://gohugo.io>Hugo</a></span>
<span class=division>|</span>
<span class=theme-info>Theme -
<a class=theme-link href=https://github.com/olOwOlo/hugo-theme-even>Even</a></span>
<span class=copyright-year>&copy;
2017 -
2024<span class=heart><i class="iconfont icon-heart"></i></span><span>Stefano Mangiola, Maria Doyle</span></span></div></footer><div class=back-to-top id=back-to-top><i class="iconfont icon-up"></i></div></div><script src=https://cdn.jsdelivr.net/npm/jquery@3.2.1/dist/jquery.min.js integrity="sha256-hwg4gsxgFZhOsEEamdOYGBf13FyQuiTwlAQgxVSNgt4=" crossorigin=anonymous></script>
<script src=https://cdn.jsdelivr.net/npm/slideout@1.0.1/dist/slideout.min.js integrity="sha256-t+zJ/g8/KXIJMjSVQdnibt4dlaDxc9zXr/9oNPeWqdg=" crossorigin=anonymous></script>
<script src=https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@3.1.20/dist/jquery.fancybox.min.js integrity="sha256-XVLffZaxoWfGUEbdzuLi7pwaUJv1cecsQJQqGLe7axY=" crossorigin=anonymous></script>
<script type=text/javascript src=/tidyomicsBlog/js/main.min.4ae89da218555efa0e7093a20b92017d2e1202b66fff9fc2edf4cb8d44b44c6e.js></script></body></html>